src/cpp/session/modules/SessionObjectExplorer.cpp

```diff
@@ -0,0 +1,226 @@
+/*
+ * SessionObjectExplorer.cpp
+ *
+ * Copyright (C) 2009-17 by RStudio, Inc.
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#define R_INTERNAL_FUNCTIONS
+
+#include "SessionObjectExplorer.hpp"
+
+#include <boost/bind.hpp>
+#include <boost/foreach.hpp>
+
+#include <core/Algorithm.hpp>
+#include <core/Error.hpp>
+#include <core/Exec.hpp>
+
+#include <r/RExec.hpp>
+#include <r/RRoutines.hpp>
+#include <r/RSexp.hpp>
+
+#include <session/SessionModuleContext.hpp>
+
+using namespace rstudio::core;
+
+namespace rstudio {
+namespace session {
+namespace modules {      
+namespace explorer {
+
+namespace {
+
+const char * const kExplorerCacheDir = "explorer-cache";
+
+FilePath explorerCacheDir() 
+{
+   return module_context::sessionScratchPath()
+         .childPath(kExplorerCacheDir);
+}
+
+std::string explorerCacheDirSystem()
+{
+   return string_utils::utf8ToSystem(explorerCacheDir().absolutePath());
+}
+
+void removeOrphanedCacheItems()
+{
+   Error error;
+   
+   // if we don't have a cache, nothing to do
+   if (!explorerCacheDir().exists())
+      return;
+   
+   // list source documents
+   std::vector<boost::shared_ptr<source_database::SourceDocument> > docs;
+   error = source_database::list(&docs);
```

Comment:
This is somewhat expensive as it has to crawl the entire source database. In this case all we need is a list of the IDs in the source database (a happy result of using the ID as the cache key!) so could simply get the list of IDs via something like `source_database::path().children(&files);` (maybe wrapped behind an accessor)